content/articles/integrate-ml-to-flask-api-and-deploy-to-heroku/index.md

```diff
@@ -0,0 +1,338 @@
+
+After sending the POST request, we get the model prediction as a response. Let us see what happens if the client sends the wrong JSON data to the API.
+
+![POST request with the wrong format to local server](/engineering-education/integrate-ml-to-flask-api-and-deploy-to-heroku/post_wrong_request_to_local_server.PNG)
+
+If we send the wrong JSON data to the API, we get an error message with the 500 status code. However, the client will also receive the correct JSON data format. Since our API is working fine, we can proceed to deploy it.
+
+
+### Deploying to Heroku
+
+The next step is to deploy our API so other developers can have access to it. Next, we will deploy the API to Heroku. Heroku is a cloud platform that allows developers to build, run, and deploy their apps. If you do not have a Heroku account, you will have to create one by clicking [here](https://signup.heroku.com/).
+
+We need to install Git and Heroku CLI (Command Line Interface) on our PC. Heroku uses git to manage the deployment of apps. Click [here](https://git-scm.com/downloads) to download and install git. Use the link [here](https://devcenter.heroku.com/articles/heroku-cli#download-and-install) to download and install Heroku CLI. 
+
+With Heroku CLI, we can use our terminal to deploy. The Heroku CLI allows us to access Heroku from our terminal. With it, we can create and manage applications on Heroku with ease. To access Heroku CLI, we log in to Heroku from our terminal. 
+
+```bash
+heroku login
+```
+
+This command takes us to the Heroku login page on our web browser. If you want to remain in the CLI and log in, use the command below instead.
+
+```bash
+heroku login -i
+```
+
+After successfully logging in, we can proceed. We can initialize a git repo in our project directory after installing git.
+
+```bash
+git init
+```
+
+We need to add two files to our project directory: `Procfile` and `requirements.txt`. The `Procfile` tells Heroku what to do with the code you want to push to their server. The `requirements.txt` contains all the dependencies we need to install to run our API on Heroku. We will be running the API using `Gunicorn`. First, we need to install it. 
+
+`Gunicorn` is a web server that is more powerful than the built-in server that flask gives you. The built-in server in the flask can only handle one user at a time. But `Gunicorn` can deal with many users at once.
+
+```bash
+pip install gunicorn
+```
+
+Create a new file, `Procfile`, and add the code below in the file. 
+
+```
+web: gunicorn api: app
+```
+Here, the `Procfile` tells Heroku that we want to run a web API using gunicorn. The api in the code above represents the name of our python script. So, we want to run the app inside the `api.py` script.
+
+For the `requirements.txt` file, run the command below in your terminal. 
+
+```bash
+pip freeze > requirements.txt
+```
+
+Running the command creates a requirements.txt. It contains all the libraries installed in the virtual environment.
+
+```

```

Comment:
The libraries and their corresponding versions from the requirements.txt file are listed here